compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +447,243 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
   return DiagnosedSilenceableFailure(success());
 }
 
+/// Return `true` if the given type is a ShapedType and has at least one
+/// dynamic dimension.
+static bool hasDynamicShape(Type t) {
+  auto shapedType = t.dyn_cast<ShapedType>();
+  if (!shapedType) return false;
+  return !shapedType.hasStaticShape();
+}
+
+/// Reify the dynamic dimensions of the given value.
+static LogicalResult reifyDynamicResultDims(OpBuilder b, Value value,
+                                            SmallVector<Value> &dynamicDims) {
+  OpBuilder::InsertionGuard guard(b);
+
+  // Case 1: No dynamic result dims.
+  if (!hasDynamicShape(value.getType())) return success();
+
+  // There is at least one dynamic dimension, continue...
+  ShapedType shapedType = value.getType().cast<ShapedType>();
+
+  // Case 2: Value is a block argument.
+  if (auto bbArg = value.dyn_cast<BlockArgument>()) {
+    b.setInsertionPointToStart(bbArg.getOwner());
+    for (int64_t i = 0; i < shapedType.getRank(); ++i) {
+      if (shapedType.isDynamicDim(i)) {
+        Value dim = b.create<tensor::DimOp>(bbArg.getLoc(), bbArg, i);
+        dynamicDims.push_back(dim);
+      }
+    }
+    return success();
+  }
+
+  // Value is an OpResult.
+  Operation *op = value.getDefiningOp();
+  OpResult opResult = value.cast<OpResult>();
+  b.setInsertionPoint(op);
+
+  // Case 3: Value is tied. Reify the dimensions of the tied operand.
+  auto tiedOp = dyn_cast<IREE::Util::TiedOpInterface>(op);
+  if (tiedOp) {
+    Value tiedOperand = tiedOp.getTiedResultOperand(value);
+    if (tiedOperand) {
+#ifndef NDEBUG
+      ShapedType tiedOperandType = tiedOperand.getType().cast<ShapedType>();
+      assert(tiedOperandType == shapedType && "expected same type");
+#endif  // NDEBUG
+      return reifyDynamicResultDims(b, tiedOperand, dynamicDims);
+    }
+  }
+
+  // Case 4: Query ShapeAwareOpInterface.
+  auto shapeAwareOp = dyn_cast<IREE::Util::ShapeAwareOpInterface>(op);
```

Comment:
Maybe better to move this to case 3 cause this would mot require any new operations.

Reply:
Do you mean moving this before Step 3? I just gave it a try and it doesn't break anything, but I think the current approach could be a bit more efficient.

The way the two new ops are used is as follows: Create a dispatch region with just one op with `WrapInDispatchRegionOp`. Then use `MoveIntoDispatchRegionOp` iteratively to pull in (clone) more and more ops into the region. The problem is what if we are trying to pull in an op that is also used in a `tensor.dim` that is used as `resultDims` operand for the `dispatch.region` op. In that case, the pulled-in op will not fold away outside of the region because it still has a use. And we have to rely on other patterns canonicalize the `tensor.dim` ops.

Example:
```mlir
%0 = "some_op"(%arg0) : tensor<?xf32>
%1 = tensor.dim %0, %c0
%r = dispatch.region -> (tensor<?xf32>{%1}) {...}
```
Now we are cloning `%0` into the region. But the original `%0` won't DCE away because of `%1`. (It will eventually fold away after the `tensor.dim` is canonicalized to `dim(arg0)`.) We can avoid this by walking up the tied OpResult-OpOperand chain as much as possible when calculating the dynamic dims.